### ISSUE TITLE
Elect a renaming handler to check new references

### ISSUE THREAD

**Issue by jvcoutinho** [guilhermejccavalcanti/jFSTMerge#124]
Currently, only one of the four renaming handlers check if a developer added a new reference to an edited method, and only under certain conditions. Although, another handler, [NewElementReferencingEditedOne](https://github.com/guilhermejccavalcanti/jFSTMerge/blob/master/src/main/java/br/ufpe/cin/mergers/handlers/NewElementReferencingEditedOneHandler.java), checks if a developer added a new element that refers to an edited one, but again, under certain conditions.

So, as solution, we could do the following:

- Elect one of the handlers to check new references. I vote for Merge Methods, once it's one of the simplest handlers and acts like an extension to semistructured merge by merging two nodes with different identifiers.
- Implement this functioning in all of the four renaming handlers.
- Improve NewElementReferencingEditedOne handler to check this more reliably.

**Comment by pauloborba** [comment]
slides: https://docs.google.com/presentation/d/15U41vG-Ljk62zXyJ3qKhExY7EYVrNVtrBJlTICpYpxs/edit?usp=sharing

[handlersalgorithm-2.pdf](https://github.com/guilhermejccavalcanti/jFSTMerge/files/3912253/handlersalgorithm-2.pdf)

**Comment by pauloborba** [comment]
@jvcoutinho we have to first discuss how precise is the reference check. and how far it requires the project to be compilable? buildable?

**Comment by jvcoutinho** [comment]
@pauloborba currently, it just checks at the same class level. This means new references to public methods in other classes aren't detected. 
We can expand this to the whole project (makes more sense) or to the package level. But in both scenarios, this would be a feature that would only make sense if S3M runs as the git driver, because you can plug files without packages in the standalone version.

**Comment by pauloborba** [comment]
@jvcoutinho does the current (class scoped reference check) implementation require the class to be syntactic valid? compilable?

@guilhermejccavalcanti given this class scope restriction, would you expect much difference from this reference check (in relation to the other handlers that do no checking)

**Comment by jvcoutinho** [comment]
> @jvcoutinho does the current (class scoped reference check) implementation require the class to be syntactic valid? compilable?

Yes. We compile the class in order to catch the references.

**Comment by guilhermejccavalcanti** [comment]
> 
> @guilhermejccavalcanti given this class scope restriction, would you expect much difference from this reference check (in relation to the other handlers that do no checking)

I don't think so, at least from the observations of the NewElementReferencingEditedOne handler, there are very few occurrences. I can't tell about the cases of renaming.

But the main limitation of the compilation scope is the fact that on real use (as a git merge driver), the tool has only access to the merged file, not the entire project.
